CIP-optimized-uplc-serialization/README.md

```diff
@@ -0,0 +1,446 @@
+- ```I``` is followed by a **signed** integer
+- ```B``` is followed by a ```ByteString```
+
+proposed ```data``` serialization (option 2):
+
+introducing specific 2 bits tags for a generic ```data``` element
+
+data tag  | binary 
+----------|--------
+data-pair | 00     
+data-list | 01     
+data-int  | 10     
+data-bs   | 11     
+
+data constructors are mapped to the above tags as
+
+data constructor | data tag  | binary 
+-----------------|-----------|--------
+Constr           | data-pair | 00     
+Map              | data-list | 01     
+List             | data-list | 01     
+I                | data-int  | 10     
+B                | data-bs   | 11     
+
+with the aim of making the serialization and deserialization of the constructors ```Map``` and ```List``` unambiguous, only the ```data-list``` tag requires an additional bit indicating the intended constructor; adding ```0``` for ```Map```s and ```1``` for ```List```s
+
+the final encoding of constructors is then:
+
+data constructor | data tag         | binary 
+-----------------|------------------|--------
+Constr           | data-pair        | 00     
+Map              | data-list (map)  | 01 0
+List             | data-list (list) | 01 1
+I                | data-int         | 10
+B                | data-bs          | 11
+
+the encoding of the value is then the same of the one specified in option 1
+
+## Rationale
+
+This proposal suggest ways to reduce serialized scripts size.
+
+the changes where designed to keep the bit oriented style of flat minimizing the number of required bits for values where possible.
+
+here the motivations behind the choices in design are made explicit:
+
+### ```pad0``` encodes as no bits
+
+sometimes it is necessary to allign the serialized program (partial or total) to fit into a byte;
+
+in order to do so, the specification includes a ```pad``` fuction used to allign a partially serialized program prior the append of an elements that requires it, and at the end of a program in order to encode the result as a ```ByteString```.
+
+by definition ```pad``` adds ```n - 1``` ```0```s directly followed by a ```1```.
+
+the needed padding can be obtained as follows
+```ts
+pad( 8 - (currBitLength % 8) )
+```
+where the ```pad``` function could be defined as follows in typescript
+```ts
+pad( missingBits: 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 ): string
+{
+    switch( missingBits )
+    {
+        case 0: return "00000001";
+        case 1: return "1";
+        case 2: return "01";
+        case 3: return "001";
+        case 4: return "0001";
+        case 5: return "00001";
+        case 6: return "000001";
+        case 7: return "0000001";
+    }
+}
+```
+the case in which the ```missingBits``` is 0 implies that the current serialized program is already byte-alligned
+
+since this padding carries no usefull informations, the current ```pad( 0 )``` adds a useless byte each time a padding is needed and the number of used bits is a multiple of 8.
```

Comment:
I think that the reason for this is so that the decoder can easily discard padding at the end without having to check whether it's at a byte boundary, and this makes it faster.  The proposed change would on average save one bit per bytestring, so I'm not sure if it's worth it.

Reply:
if the proposed changes are accepted the only place where padding will be used is at the end of the script; since bytestring will no longer require to be byte alligned

the check for ```pad( 0 )``` then becomes just ```lengthInBits `mod` 8 == 0```